IoHandle.md

```diff
@@ -0,0 +1,387 @@
+
+### Refactoring the FD
+
+
+##### Overview
+
+The issue with refactoring the fd is the implicit assumption that the main interface to Envoy from the application is a socket. In this PR, our purpose is to abstract away sockets and all system calls pertaining to sockets from the I/O interface. The IoHandle represents the abstract interface for all I/O operations and the details of any particular interface (socket, VPP, etc) are contained in classes derived from IoHandle. 
+To this end, we also define the IoAddress class and some derivatives, as well as the IoDescriptor class, which will replace the fd() calls when dealing with actual sockets.
+
+These changes in this PR are a proposal and are intended to get feedback for the direction we are taking. Rather than show all the changes required to implement this new direction, we will also discuss towards the end how we anticipate further changes rippling through the code.
+
+Note also that this PR does not address events yet, as previous comments suggested. The intention is to introduce further PRs that will address events and the dispatcher.
+
+
+
+###### IoHandle and derivative classes
+
+The following definitions would be contained in a new file io_defs.h  (envoy/include/envoy/network):
+
+```
+
+// Generalize the SysCallResult to accommodate I/O operations that are not
+// carried out by the kernel stack, hence not accomplished through a syscall
+
+template <typename T> struct IoCallResult : public struct IoResult  { 
+      T rc_;
+       int errno_;
+};
+typedef IoCallResult<int> IoCallIntResult;
+typedef IoCallResult<ssize_t> IoCallSizeResult;
+typedef IoCallResult<void*> IoCallPtrResult;
+
+template <typename T> class IoDescriptor  {
+public:
+   IoDescriptor(T descriptor):descriptor_(descriptor) { }
+   virtual ~IoDescriptor() { };
+
+   T descriptor() const { return(descriptor_); }
+
+   operator(T) const { return(descriptor_);
+   bool	operator == (const T& lhs) { return(descriptor_== lhs); }
+   bool	operator != (const T& lhs) { return(descriptor_!=lhs); }
+
+protected:
+   T descriptor_;
+};
+
+typedef IoDescriptor<int> IoSocketDescriptor;  // fd for sockets
+typedef IoDescriptor<uint32_t> IoVppDescriptor;  // for example…
+
+// I/O options
+
+struct IoOptionName { };
+struct IoOptionValue { };
+
+struct IoOptionNameSimple : public IoOptionName  { 
+     std::string name_;
+ };
+struct IoOptionValueSimple  : public IoOptionValue { 
+     std::string  value_;
+ };
+
+ struct IoSocketOptionName : public IoOptionName  { 
+    Int  level_;  
+    int  name_;
+ };
+
+ struct IoSocketOptionValue  : public IoOptionValue { 
+    int value_;
+ };
+
+class IoHandle { 
+public:
+   IoHandle() { } 
+   virtual ~IoHandle() { }
+
+   // I/O Options API
+   virtual IoCallResult setOption(IoOptionName name, IoOptionValue value) PURE;
```

Comment:
If I understand correctly, the `setOption` and the `option` methods on this `IoHandle` class are meant to correspond to `setsockopt` and `getsockopt` in the socket case?

If so, then the options supported by `setsockopt` / `getsockopt` are broadly the same. However we have found a couple options that are only supported on one platform or the other (e.g. `SO_ORIGINAL_DST` on Linux or `SO_PROTOCOL_INFO` on Windows).

Our initial impression is that this should be mostly fine, though of course we can't know for sure until we start actually implementing this interface

Reply:
My feeling here is the same in that I wonder if it would be better to actually have a specific call for every option we want to set. E.g., setNoDelay(), etc. I realize today we support arbitrary setsockopt options in the config via allowing users to specify integers, and that will have to be reconciled anyway for windows, so maybe we can consider that? But in general as we move to more platforms IMO it's much better to be explicit about the intention of each option we want to set?